### PR TITLE
Add setting config

### PR DESCRIPTION
**Pull request by dankmitchell**
### Context
Application settings

### Changes proposed in this pull request
- Add config gem
- Replace environment variables with settings
- Remove dot-env gem

### Guidance to review
No difference

Note: before we merge to prod we need to add `SETTINGS__DFE_SIGNIN__SECRET`. Then once deployed remove the following -

```
DFE_SIGN_IN_ISSUER
DFE_SIGN_IN_IDENTIFIER
DFE_SIGN_IN_SECRET
DFE_SIGN_IN_PROFILE
BASE_URL
MANAGE_BACKEND_BASE_URL
MANAGE_UI_BASE_URL
```

### CODE REVIEW

**Comment by dankmitchell** [comment]
[Add config](https://trello.com/c/wMhXPmps/1060-add-config)

File: config/initializers/config.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
+Config.setup do |config|
+  # Name of the constant exposing loaded settings
+  config.const_name = 'Settings'
```
Comment (defong) [inline config/initializers/config.rb]: https://github.com/DFE-Digital/manage-courses-backend/blob/214f6da771de448408b8c0b2aa24eea9945a7c83/config/initializers/config.rb#L24

File: spec/controllers/sessions_controller_spec.rb [hunk 2]
```diff
@@ -11,7 +11,7 @@
 
       get :signout
 
-      expect(subject).to redirect_to("https://example.com/session/end?id_token_hint=123&post_logout_redirect_uri=https://localhost:3000/auth/dfe/signout")
+      expect(subject).to redirect_to("https://signin-test-oidc-as.azurewebsites.net/session/end?id_token_hint=123&post_logout_redirect_uri=https://localhost:3000/auth/dfe/signout")
```
Comment (defong) [inline spec/controllers/sessions_controller_spec.rb]: hardcoding

File: config/settings.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+dfe_signin:
+  identifier: bats2
```
Comment (defong) [inline config/settings.yml]: not too sure should it be in repo...

File: config/initializers/config.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+  # Define ENV variable prefix deciding which variables to load into config.
+  #
+  config.env_prefix = 'SETTINGS'
+
+  # What string to use as level separator for settings loaded from ENV variables. Default value of '.' works well
+  # with Heroku, but you might want to change it for example for '__' to easy override settings from command line, where
+  # using dots in variable names might not be allowed (eg. Bash).
+  #
+  # config.env_separator = '.'
```
Comment (defong) [inline config/initializers/config.rb]: https://github.com/DFE-Digital/manage-courses-backend/blob/214f6da771de448408b8c0b2aa24eea9945a7c83/config/initializers/config.rb#L30

File: config/initializers/config.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+  # Ability to process variables names:
+  #   * nil  - no change
+  #   * :downcase - convert to lower case
+  #
+  # config.env_converter = :downcase
+
+  # Parse numeric values as integers instead of strings.
+  #
+  # config.env_parse_values = true
```
Comment (defong) [inline config/initializers/config.rb]: https://github.com/DFE-Digital/manage-courses-backend/blob/214f6da771de448408b8c0b2aa24eea9945a7c83/config/initializers/config.rb#L40

File: spec/controllers/sessions_controller_spec.rb (hunk 2 above)
Comment (dankmitchell) [reply spec/controllers/sessions_controller_spec.rb]: This was intentional...

File: config/settings.yml (hunk 3 above)
Comment (dankmitchell) [reply config/settings.yml]: Was intentional. It's exposed through the callback URL to the client

File: azure/template.json [hunk 6]
```diff
@@ -247,6 +247,10 @@
                                 "name": "DFE_SIGN_IN_SECRET",
                                 "value": "[parameters('dfeSignInSecret')]"
                             },
+                            {
+                                "name": "SETTINGS__DFE__SIGN__IN__SECRET",
+                                "value": "[parameters('dfeSignInSecret')]"
+                            },
                             {
                                 "name": "DFE_SIGN_IN_PROFILE",
```
Comment (defong) [inline azure/template.json]: Can i reject based on you added "SETTINGS__DFE__SIGN__IN__SECRET"

and not update this file?

Comment (defong) [reply azure/template.json]: feeling the need to reject you as deployment to via azure template will be broken

**Comment by defong** [comment]
Since the gem is gone can we also remove https://github.com/DFE-Digital/manage-courses-frontend/blob/6e9944dc1426a367eefc631af89d27aa0160ba07/docker-compose.yaml#L12

File: azure/template.json [hunk 7]
```diff
@@ -247,6 +247,10 @@
                                 "name": "DFE_SIGN_IN_SECRET",
                                 "value": "[parameters('dfeSignInSecret')]"
                             },
+                            {
+                                "name": "SETTINGS__DFE__SIGN__IN__SECRET",
```
Comment (misaka) [inline azure/template.json]: Single underscore: `...DFE_SIGNIN...`.

File: config/initializers/config.rb (hunk 1 above)
Comment (misaka) [reply config/initializers/config.rb]: Wrong var name, it's `const_name` not `env_prefix`.

**Pull request merged**